### ISSUE TITLE
Optimization for when entire contract has no code

### ISSUE THREAD

**Issue by fulldecent** [ethereum/solidity#12496]
## Abstract

A few, specific contracts can include a specific optimization.

This helps a class of people that want to use Yul.

## Motivation

To make Yul easy to deploy.

## Test case

This contract:

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.11;

contract ERC20 {
    fallback () external payable {
    }
}
```

Should deploy with just the hash at the end. But instead it deploy with all this boilerplate:

```
        0000    60  PUSH1 0x80
	0002    60  PUSH1 0x40
	0004    52  MSTORE
	0005    00  *STOP

	0006    FE    *ASSERT
	0007    A2    LOG2
	0008    64    PUSH5 0x6970667358
	000E    22    22
	000F    12    SLT
	0010    20    SHA3
	0011    8B    DUP12
	0012    00    *STOP
	0013    C3    C3
	0014    09    MULMOD
	0015    32    ORIGIN
	0016    9C    SWAP13
	0017    D3    D3
	0018    92    SWAP3
	0019    76    PUSH23 0x14e702eaae5d8fc4c88db71c4fa005baea66b0c202acb1
	0031    64    PUSH5 0x736f6c6343
	0037    00    *STOP
	0038    08    ADDMOD
	0039    0B    SIGNEXTEND
	003A    00    *STOP
	003B    33    CALLER
```

## More motivation

Why would anybody want to deploy that exact contract? And do you really care about removing the three extraneous instructions at top?

Yes.

Here's why.

This is actually the best (only) for most people to deploy a Yul contract. For example, deploying [the Solidity documentation's ERC-20 contract with Yul](https://docs.soliditylang.org/en/v0.8.11/yul.html?highlight=yul#complete-erc20-example) can be done with:

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.11;

contract ERC20 {
    fallback () external payable {
        assembly {
           ... paste the ERC-20 Yul code here...
        }
    }
}
```

The kind of people that would deploy Yul are the kind of people that care about these three extra instructions.

Making the proposed change above would allow everybody that has access to a Solidity toolchain to deploy Yul code with zero overhead. For example, Hardhat, Truffle, and basically everything else does not (built in) support Yul. But with this adapter it does.

## Specification

When Solidity is compiling a contract with no usage of memory, do not synthesize the memory size storage location.

I.e., don't include this:

```
        0000    60  PUSH1 0x80
	0002    60  PUSH1 0x40
	0004    52  MSTORE
```

Separately, in the documentation for Yul, include the above code as the recommended and fully efficient way to deploy Yul code.

## Backwards Compatibility

No considerations

**Comment by hrkrshnn** [comment]
We are working on an optimization step that would remove unnecessary `mstores`: https://github.com/ethereum/solidity/pull/11352  that should fix the issue, for the (currently experimental) Yul based compilation pipeline.

**Comment by chriseth** [comment]
I agree, this will be solved by #11352 and I don't see a big benefit in implementing this in the opcode-based optimizer, since the cited PR for the yul optimizer can do so much more than just this special case.

**Comment by haltman-at** [comment]
> For example, Hardhat, Truffle, and basically everything else does not (built in) support Yul.

FWIW, Truffle employee popping in here to point out that Truffle does now have rudimentary Yul support. :)
